### PR TITLE
8293331: Refactor FileDispatcherImpl into operating system-specific components

### PR DESCRIPTION
**Pull request by bplb**
Break up `sun.nio.ch.FileChannelImpl` in a manner similar but not identical to the refactoring of `sun.nio.fs.UnixFileSystem` performed in #10093.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293331](https://bugs.openjdk.org/browse/JDK-8293331): Refactor FileDispatcherImpl into operating system-specific components

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10154/head:pull/10154` \
`$ git checkout pull/10154`

Update a local copy of the PR: \
`$ git checkout pull/10154` \
`$ git pull https://git.openjdk.org/jdk pull/10154/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10154`

View PR using the GUI difftool: \
`$ git pr show -t 10154`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10154.diff">https://git.openjdk.org/jdk/pull/10154.diff</a>

</details>

### CODE REVIEW

**Comment by bplb** [comment]
The initial implementation moves most of `FileChannelImpl` to `sun.nio.ch.AbstractFileChannelImpl` and requires separate but simple `FileChannelImpl.java` and `FileChannelImpl.c` files for each operating system. The AIX, Linux, and macOS versions extend an intermediate class `sun.nio.ch.UnixFileChannelImpl` which itself extends `AbstractFileChannelImpl`.

All tests in the `jdk_nio` test group pass on Linux, macOS, and Windows.

Although an AIX implementation is provided, it will have to be verified at a later date by someone with access to the requisite platform.

**Comment by AlanBateman** [comment]
FileChannel delegates to a FileDispatcher so there is already a place for OS specific code (both java and native code). I think we should try to lean on that before introducing a parallel hierarchy of classes.

**Comment by bplb** [comment]
> FileChannel delegates to a FileDispatcher so there is already a place for OS specific code (both java and native code). I think we should try to lean on that before introducing a parallel hierarchy of classes.

Having thought more about it over the long weekend I don't like this proposal as it stands either. The native methods as-is however I thought are too long for the usual `FileDispatcher` standard. I will however investigate this alternative.

**Comment by AlanBateman** [comment]
I took a first pass over the latest update and it looks quite good. I assume the issue should be changed now as this is about refactoring of FileChannelImpl rater than FileChannelImpl.

Could FileChannelImpl.nd be a static field, there doesn't seen to be any reason to create one per FileChannel now. Also Unmapper shouldn't need its own now either.

Can the static initialiser that calls IOUtil.load be removed from FileChannelImpl? The library loading is pushed down to FileDispatcherImpl now so I assume it is not needed.

**Comment by bplb** [comment]
> Could FileChannelImpl.nd be a static field, there doesn't seen to be any reason to create one per FileChannel now. Also Unmapper shouldn't need its own now either.
> 
> Can the static initialiser that calls IOUtil.load be removed from FileChannelImpl? The library loading is pushed down to FileDispatcherImpl now so I assume it is not needed.

Above all fixed in c6e271eb0d23f846d4c5ef55c3cb32ba2f7d896d.

**Comment by AlanBateman** [comment]
> Although an AIX implementation is provided, it will have to be verified at a later date by someone with access to the requisite platform.

@backwaterred Are you able to test the AIX changes?

**Review by AlanBateman — APPROVED** [review]
The overall refactoring looks okay. Most of the change is code moving to another file so it's hard to see if there are any changes and it's hard to see if the list of include files is accurate. Everything I checked seemed okay and unchanged.

**Comment by backwaterred** [comment]
I am happy to test this on AIX @bplb & @AlanBateman. I will update this message when it completes.

---

Testing in progress

**Comment by bplb** [comment]
> I am happy to test this on AIX @bplb & @AlanBateman. I will update this message when it completes.
> 
> Testing in progress

Thank you!

**Comment by bplb** [comment]
> test/jdk/jdk/nio... 27/27 passing

@backwaterred If tests passed then there was no build issue so I will integrate this tomorrow. Thanks for testing!

**Comment by bplb** [comment]
> [...] it's hard to see if the list of include files is accurate.

Unneeded include files are removed in a2da4deba931fdb67e5bd8b5c34bb562490ace4e. I don't know whether the absence of those removed in `UnixFileDispactherImpl` will cause problems on AIX (@backwaterred please test the build again, thanks).

**Comment by AlanBateman** [comment]
The update removed the include of jni.h but JNI types are used. They get included by the javac generated header files but might be clearer to not remove it.

**Comment by backwaterred** [comment]
The tests are in progress :-)

**Comment by backwaterred** [comment]
Looking good @bplb. The jdk/nio tests are all passing.

**Comment by backwaterred** [comment]
> Thank you, @backwaterred, for the assist.

You're welcome :smile:

File: src/java.base/linux/classes/sun/nio/ch/FileDispatcherImpl.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    long transferFrom(FileDescriptor src, FileDescriptor dst,
+                      long position, long count, boolean append) {
+        return transferFrom0(src, dst, position, count, append);
+    }
+
+    // -- Native methods --
+
+    static native long transferTo0(FileDescriptor src, long position,
```
Comment (vyommani) [inline src/java.base/linux/classes/sun/nio/ch/FileDispatcherImpl.java:51]: can we make these native methods(transferTo0, transferFrom0 etc) to private access specifier,  these methods are not called from outside ?.

Comment (bplb) [reply src/java.base/linux/classes/sun/nio/ch/FileDispatcherImpl.java:51]: This comment applies to a lot of code. See `UnixFileDispatcherImpl` for example. I prefer not to change it at this time.

**Pull request closed without merging**